miner/agent.go

```diff
@@ -17,56 +17,49 @@
 package miner
 
 import (
-	"fmt"
 	"sync"
 
 	"sync/atomic"
 
-	"github.com/ethereum/go-ethereum/common"
-	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/consensus"
 	"github.com/ethereum/go-ethereum/log"
-	"github.com/ethereum/go-ethereum/pow"
 )
 
 type CpuAgent struct {
 	mu sync.Mutex
 
 	workCh        chan *Work
-	quit          chan struct{}
+	stop          chan struct{}
 	quitCurrentOp chan struct{}
 	returnCh      chan<- *Result
 
-	index int
-	pow   pow.PoW
+	chain  consensus.ChainReader
+	engine consensus.Engine
 
 	isMining int32 // isMining indicates whether the agent is currently mining
 }
 
-func NewCpuAgent(index int, pow pow.PoW) *CpuAgent {
+func NewCpuAgent(chain consensus.ChainReader, engine consensus.Engine) *CpuAgent {
 	miner := &CpuAgent{
-		pow:    pow,
-		index:  index,
-		quit:   make(chan struct{}),
+		chain:  chain,
+		engine: engine,
+		stop:   make(chan struct{}, 1),
 		workCh: make(chan *Work, 1),
 	}
-
 	return miner
 }
 
 func (self *CpuAgent) Work() chan<- *Work            { return self.workCh }
-func (self *CpuAgent) Pow() pow.PoW                  { return self.pow }
 func (self *CpuAgent) SetReturnCh(ch chan<- *Result) { self.returnCh = ch }
 
 func (self *CpuAgent) Stop() {
-	close(self.quit)
+	self.stop <- struct{}{}
```

Comment:
The CPU agent is reused between works. Recreating channels is very dangerous as it's extremely easy to produce data races. Previously we created new miners wily nilly (exactly because it wasn't safe to recreate the channel alone within a miner), which lead to a lot of issues. By moving threading into `ethash`, the mining code is not clean enough not to require constant recreations.